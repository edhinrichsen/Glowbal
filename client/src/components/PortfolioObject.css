.portfolio-object {
  display: inline-block;
  width: 768px;
  height: 100vh;
  transition: filter 0.3s, border-color 0.5s, width 1s, height 1s 0s,
    padding 0.3s, margin 0.3s, opacity 0.2s;
}

.thumbnail {
  padding: 0;
  height: 180px;
  width: 240px;
}



/*===========================
  portfolio exit button
  ===========================*/

.portfolio-exit {
  width: 24px;
  height: 24px;
  background-color: rgba(168, 168, 168, 0.5);
  border-radius: 24px;
  z-index: 1;
  position: relative;
  float: right;
}
.portfolio-exit::before {
  content: "âœ•";
  font-size: 1.15em;
  /* margin-top: 100px; */
}

.thumbnail .portfolio-exit::before {
  content: "";
}

.thumbnail .portfolio-exit {
  top: 0;
  right: 0;
  border: 0;
  width: 100%;
  height: 100%;
  border-radius: 0;
  background-color: rgba(0, 0, 0, 0);
}



/*===========================
  portfolio hero image
  ===========================*/

.portfolio-hero {
  background-image: url(../placeholder-folio-image.png);
  background-size: cover;
  background-position: center;
  width: 100%;
  height: 330px;

  transition: box-shadow 0.5s, border-color 0.5s, width 1s, height 0.5s 0s,
    margin 0.5s, opacity 0.2s;
}
.thumbnail .portfolio-hero {
  width: 100%;
  height: 100%;
}



/*===========================
  portfolio overlay on hover
  ===========================*/

.portfolio-object-overlay {
  opacity: 0;
  height: 0;
  font-size: 1em;
  display: flex;
  position: relative;
  top: 0%;
  padding-top: 20px;

  transition: opacity 0.3s ease, top 0.3s ease, background 0.3s ease;
  background: linear-gradient(
    180deg,
    rgba(240, 240, 240, 0) 0%,
    rgba(240, 240, 240, 0) 100%
  );
}

.thumbnail .portfolio-object-overlay {
  opacity: 0;
  top: -80%;
  background: linear-gradient(180deg, rgba(240, 240, 240, 0) 0%, #ffffff 100%);
  height: 100%;
  width: 100%;
}

.thumbnail:hover > .portfolio-object-overlay {
  opacity: 1;
  top: -100%;
}

.thumbnail:hover {
  filter: drop-shadow(0px 2px 16px rgba(0, 0, 0, 0.2));
}

.portfolio-object-overlay .portfolio-meta {
  margin-top: auto;
  text-align: left;

  border-radius: 10px;

  padding-left: 20px;
  padding-bottom: 20px;
}



/*===========================
  portfolio body and info
  ===========================*/

.portfolio-object-body {
  opacity: 1;
  transition: opacity 1s;
  text-align: left;
}
.thumbnail .portfolio-object-body {
  opacity: 0;
}



/*===========================
  portfolio title and
  author info
  ===========================*/

.portfolio-title {
  opacity: 1;
  transition: opacity 1s, height 1s, margin 1s;
  height: 1em;
  margin-bottom: 30px;
  /* background-color: aqua; */
}
.thumbnail .portfolio-title {
  opacity: 0;
  height: 0;
  margin: 0;
}

.portfolio-author {
  background-color: aqua;
  float: left;
  text-align: left;
  display: inline-flex;
}



/*===========================
  portfolio side bar
  ===========================*/

.portfolio-side-bar {
  float: right;
  text-align: left;
  width: 200px;
  height: 100%;
  margin-left: 30px;
}
.portfolio-side-bar ul {
  padding: 0;
  list-style-type: none;
}
.portfolio-side-bar li {
  padding-top: 10px;
  padding-bottom: 10px;
}



/*===========================
  hide portfolio entry when
  other entry is open
  ===========================*/

.portfolio-object-open .thumbnail {
  opacity: 0;
  width: 0;
  height: 0;
  padding: 0;
  margin: 0;
  transition: width 0.8s, height 1s 0s, padding 0.3s, margin 0.3s, opacity 0.2s;
}


