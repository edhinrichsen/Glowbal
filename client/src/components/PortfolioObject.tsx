import React, { useEffect } from 'react';
import { useState } from 'react';

import SocialLinks from './SocialLinks';
import PortfolioHero from './PortfolioHero';

import './PortfolioObject.css';
import Emoji from './Emoji';
import AuthorBadge from './AuthorBadge';
// import { useAuth0 } from '@auth0/auth0-react';
import TextInput from './TextInput';
import { useRef } from 'react';
import { DatePicker, Dropdown, Select, Tooltip } from 'antd';
import moment from 'moment';
import {
  CalendarOutlined,
  EyeOutlined,
  LineOutlined,
  LinkOutlined,
} from '@ant-design/icons';

const { Option } = Select;

type PortfolioObjectData = {
  projectID?: string;
  tags?: string[];
  views?: number;
  datePosted?: string;
  location?: string;
  projectName?: string;
  link?: string;
  userID?: string;
  projectType?: string;
};

type PortfolioObjectProps = {
  data: PortfolioObjectData;
  portfolioObjectOpen: any;
  isMyProfile: boolean;
  new: boolean;
  setData: (d: PortfolioObjectData) => void;
  delData: (d: PortfolioObjectData) => void;
};

function PortfolioObject(props: PortfolioObjectProps) {
  // const { user, isAuthenticated, isLoading } = useAuth0();

  const containerPrimaryRef = useRef<any>(null);

  const [newEntry, setNewEntry] = useState(props.new);

  const [thumbnail, setThumbnail] = useState(true);

  const date = (d?: string): string | undefined => {
    const newDate = d?.split('-');
    if (d && newDate?.length === 3) {
      return `${newDate[2].replace(/T.*/, '')}-${newDate[1]}-${newDate[0]}`;
    } else return d;
  };

  const transition = () => {
    setThumbnail(!thumbnail);
    props.portfolioObjectOpen(thumbnail);
    if (containerPrimaryRef.current !== null) {
      containerPrimaryRef.current.scrollTop = 0;
    }
  };

  useEffect(() => {
    if (props.new) {
      transition();
    }
    if (containerPrimaryRef.current !== null) {
      containerPrimaryRef.current.scrollTop = 0;
    }
  }, []);

  const formatLink = (link: string) => {
    const l = link.replace(/https?:\/\/(www.)?/, '').replace(/\/$/, '');
    return (
      <div
        style={{
          marginBottom: '-5px',
          display: 'inline-block',
          width: '140px',
          overflow: 'hidden',
          textOverflow: 'ellipsis',
        }}
      >
        {l}
      </div>
    );
  };

  return (
    <div
      ref={containerPrimaryRef}
      className={`container-primary portfolio-object ${
        thumbnail ? 'thumbnail' : 'container-scroll'
      }`}
    >
      {/* {props.data.projectID} */}
      {props.isMyProfile ? (
        <div
          onClick={() => {
            transition();
            props.delData(props.data);
          }}
          className="del-button"
        ></div>
      ) : (
        <></>
      )}

      <div
        onClick={() => {
          // if (props.data.new) {
          //   const newData = { ...props.data };
          //   newData.new = false;
          //   props.setData(newData);
          // }
          transition();
          setNewEntry(false);
        }}
        className={`${newEntry ? 'save-button' : 'exit-button'}`}
        // className={`${'exit-button'}`}
      ></div>

      <div className="portfolio-title">
        <h2 style={{ marginRight: '85px', marginLeft: '85px' }}>
          <TextInput
            padding="2px 85px 2px 85px"
            editable={props.isMyProfile}
            onChange={(newString: string) => {
              const newData = { ...props.data };
              newData.projectName = newString;
              props.setData(newData);
            }}
            text={props.data.projectName}
            placeholder="Title"
          />
        </h2>
      </div>

      <PortfolioHero
        new={props.new}
        projectID={props.data.projectID}
        isOpen={!thumbnail}
        isMyProfile={props.isMyProfile}
      />
      <div className="portfolio-object-overlay">
        <div className="portfolio-meta">
          <h4>
            <strong>{props.data.projectName}</strong>
          </h4>
          {/* <h5>{props.data.author}</h5> */}
        </div>
      </div>
      <div className="portfolio-object-body">
        <div>
          {/* <AuthorBadge author={props.data.author} tagline="Author Tagline" /> */}
          {/* <AuthorBadge
            author={props.data.author}
            tagline="Computer Science Student at the University of Melbourne"
          /> */}
          <br />
          <div style={{ marginLeft: '-10px', marginTop: '-10px' }}>
            <TextInput
              padding="10px"
              multiline={true}
              editable={props.isMyProfile}
              onChange={(newString: string) => {
                const newData = { ...props.data };
                newData.location = newString;
                props.setData(newData);
              }}
              text={props.data.location}
              placeholder="Description"
            />
          </div>
        </div>

        <div className="container-secondary portfolio-side-bar">
          <ul>
            <li>
              {props.isMyProfile ? (
                <Select
                  style={{ width: '100%' }}
                  placeholder="Project Type"
                  onChange={(value) => {
                    const newData = { ...props.data };
                    newData.projectType = value;
                    props.setData(newData);
                  }}
                  value={props.data.projectType}
                  filterOption={true}
                >
                  <Option value="website">website</Option>
                  <Option value="app">app</Option>
                  <Option value="code">code</Option>
                  <Option value="academic">academic</Option>
                  <Option value="model">model</Option>
                  <Option value="game">game</Option>
                  <Option value="video">video</Option>
                  <Option value="audio">audio</Option>
                  <Option value="presentation">presentation</Option>
                  {/* value={props.data.type} */}
                </Select>
              ) : (
                <div
                  style={{ width: '100%', textAlign: 'center' }}
                  className="portfolio-tag"
                >
                  {props.data.projectType}
                </div>
              )}
            </li>
            <li>
              {props.isMyProfile ? (
                <Select
                  mode="tags"
                  style={{ display: 'inline-block', width: '100%' }}
                  placeholder="Tags"
                  onChange={(value) => {
                    // console.log(value);
                    const newData = { ...props.data };
                    newData.tags = value;
                    props.setData(newData);
                  }}
                  value={props.data.tags}
                >
                  <Option value="React">React</Option>
                  <Option value="TypeScript">TypeScript</Option>
                  <Option value="Express">Express</Option>
                </Select>
              ) : props.data.tags ? (
                <div style={{ display: 'inline-block', width: '100%' }}>
                  {props.data.tags.map((value, index) => {
                    return <div className="portfolio-tag">{value}</div>;
                  })}
                </div>
              ) : (
                <></>
              )}
            </li>
            <li style={{ textAlign: 'center' }}>
              <LineOutlined />
            </li>
            <li>
              <CalendarOutlined />

              {' ' + date(props.data.datePosted)}
            </li>
            <li>
              <LinkOutlined />{' '}
              {props.isMyProfile ? (
                <div style={{ display: 'inline-block', width: '145px' }}>
                  <TextInput
                    editable={props.isMyProfile}
                    placeholder={'Link to project'}
                    text={props.data.link}
                    // placeholder="Link"
                    onChange={(newString: string) => {
                      const newData = { ...props.data };
                      newData.link = newString;
                      props.setData(newData);
                    }}
                  />
                </div>
              ) : (
                <a target="_blank" href={props.data.link}>
                  {props.data.link ? formatLink(props.data.link) : null}
                </a>
              )}
            </li>
            {/* <li>
                  <EyeOutlined /> {props.data.views}
                </li> */}
          </ul>
          <div style={{ textAlign: 'center' }}></div>
        </div>
      </div>
    </div>
  );
}

export default PortfolioObject;
