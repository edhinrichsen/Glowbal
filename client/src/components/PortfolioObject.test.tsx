import React from 'react';
import { shallow } from 'enzyme';
import PortfolioObject from './PortfolioObject';

describe('Testing PortfolioObject', () => {
  /* TODO: update when API is linked */

  it('renders without crashing', () => {
    const m = [
      { type: 'image', url: 'https://i.ibb.co/BNZxQ2z/example0.jpg' },
      { type: 'image', url: 'https://i.ibb.co/TYYyXDH/example1.png' },
      { type: 'image', url: 'https://i.ibb.co/pZmXQb5/example2.png' },
      { type: 'image', url: 'https://i.ibb.co/SwzRr9S/example3.png' },
      {
        type: 'pdf',
        url:
          'https://www.w3.org/WAI/ER/tests/xhtml/testfiles/resources/pdf/dummy.pdf',
      },
      {
        type: 'video',
        url:
          'http://distribution.bbb3d.renderfarming.net/video/mp4/bbb_sunflower_1080p_30fps_normal.mp4',
      },
    ];
    const component = shallow(
      <PortfolioObject
        id="1"
        title="Project 1"
        type="website"
        // picture="https://i.ibb.co/BNZxQ2z/example0.jpg"
        date="2020-01"
        author="Author 1"
        shortDescription="This is a short description about the project"
        views="10"
        location="Melbourne, Australia"
        media={m}
        link="http://distribution.bbb3d.renderfarming.net/video/mp4/bbb_sunflower_1080p_30fps_normal.mp4"
        portfolioObjectOpen={false}
      />
    );
    expect(component).toMatchSnapshot();
    component.unmount();
  });
});
