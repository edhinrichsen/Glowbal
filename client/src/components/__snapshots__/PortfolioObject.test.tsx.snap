// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing PortfolioObject renders without crashing 1`] = `
<div
  className="container-primary portfolio-object thumbnail"
>
  <div
    className="exit-button"
    onClick={[Function]}
  />
  <div
    className="portfolio-title"
  >
    <h2
      style={
        Object {
          "marginLeft": "85px",
          "marginRight": "85px",
        }
      }
    >
      <TextInput
        editable={false}
        onChange={[Function]}
        padding="2px 85px 2px 85px"
        placeholder="Title"
        text="Project 1"
      />
    </h2>
  </div>
  <PortfolioHero
    isMyProfile={false}
    isOpen={false}
    new={false}
    projectID="1"
    userID="10"
  />
  <div
    className="portfolio-object-overlay"
  >
    <div
      className="portfolio-meta"
    >
      <h4>
        <strong>
          Project 1
        </strong>
      </h4>
    </div>
  </div>
  <div
    className="portfolio-object-body"
  >
    <div>
      <br />
      <div
        style={
          Object {
            "marginLeft": "-10px",
            "marginTop": "-10px",
          }
        }
      >
        <TextInput
          editable={false}
          multiline={true}
          onChange={[Function]}
          padding="10px"
          placeholder="Description"
          text="Melbourne, Australia"
        />
      </div>
    </div>
    <div
      className="container-secondary portfolio-side-bar"
    >
      <ul>
        <li>
          <div
            className="portfolio-tag"
            style={
              Object {
                "textAlign": "center",
                "width": "100%",
              }
            }
          >
            website
          </div>
        </li>
        <li>
          <div
            style={
              Object {
                "display": "inline-block",
                "width": "100%",
              }
            }
          >
            <div
              className="portfolio-tag"
            >
              one
            </div>
            <div
              className="portfolio-tag"
            >
              two
            </div>
            <div
              className="portfolio-tag"
            >
              three
            </div>
            <div
              className="portfolio-tag"
            >
              four
            </div>
            <div
              className="portfolio-tag"
            >
              five
            </div>
          </div>
        </li>
        <li
          style={
            Object {
              "textAlign": "center",
            }
          }
        >
          <ForwardRef(LineOutlined) />
        </li>
        <li>
          <ForwardRef(CalendarOutlined) />
           2020-01
        </li>
        <li>
          <ForwardRef(LinkOutlined) />
           
          <a
            href="http://distribution.bbb3d.renderfarming.net/video/mp4/bbb_sunflower_1080p_30fps_normal.mp4"
            rel="noopener noreferrer"
            target="_blank"
          >
            <div
              style={
                Object {
                  "display": "inline-block",
                  "marginBottom": "-5px",
                  "overflow": "hidden",
                  "textOverflow": "ellipsis",
                  "width": "140px",
                }
              }
            >
              distribution.bbb3d.renderfarming.net/video/mp4/bbb_sunflower_1080p_30fps_normal.mp4
            </div>
          </a>
        </li>
      </ul>
      <div
        style={
          Object {
            "textAlign": "center",
          }
        }
      />
    </div>
  </div>
</div>
`;
